testscript/exe.go

```diff
@@ -49,61 +50,116 @@ func IgnoreMissedCoverage() {
-			if err := finalizeCoverProfile(); err != nil {
-				log.Printf("cannot merge cover profiles: %v", err)
+			if err := os.RemoveAll(bindir); err != nil {
+				log.Printf("cannot delete bin directory: %v", err)
 				exitCode = 2
 			}
 		}()
+		os.Setenv("PATH", bindir+string(filepath.ListSeparator)+os.Getenv("PATH"))
+
+		flag.Parse()
+		// If we are collecting a coverage profile, set up a shared
+		// directory for all executed test binary sub-processes to write
+		// their profiles to. Before finishing, we'll merge all of those
+		// profiles into the main profile.
+		if coverProfile() != "" {
+			// TODO: same as the points for bindir
+			coverdir, err := ioutil.TempDir("", "testscript-cover")
+			if err != nil {
+				log.Printf("could not set up cover directory: %v", err)
+				return 2
+			}
+			os.Setenv("TESTSCRIPT_COVER_DIR", coverdir)
+			defer func() {
+				if err := finalizeCoverProfile(coverdir); err != nil {
+					log.Printf("cannot merge cover profiles: %v", err)
+					exitCode = 2
+				}
+			}()
+		}
+
 		// We're not in a subcommand.
 		for name := range commands {
 			name := name
+			// Set up this command in the directory we added to $PATH.
+			binfile := filepath.Join(bindir, name)
+			if runtime.GOOS == "windows" {
+				binfile += ".exe"
+			}
+			if err := linkFile(os.Args[0], binfile); err != nil {
+				log.Printf("could not set up %s in $PATH: %v", name, err)
+				return 2
+			}
 			scriptCmds[name] = func(ts *TestScript, neg bool, args []string) {
-				path, err := os.Executable()
-				if err != nil {
-					ts.Fatalf("cannot determine path to test binary: %v", err)
-				}
-				id := atomic.AddInt32(&profileId, 1) - 1
-				oldEnvLen := len(ts.env)
-				cprof := coverFilename(id)
-				ts.env = append(ts.env,
-					"TESTSCRIPT_COMMAND="+name,
-					"TESTSCRIPT_COVERPROFILE="+cprof,
-				)
-				ts.cmdExec(neg, append([]string{path}, args...))
-				ts.env = ts.env[0:oldEnvLen]
-				if cprof == "" {
-					return
-				}
-				f, err := os.Open(cprof)
-				if err != nil {
-					if ignoreMissedCoverage {
-						return
-					}
-					ts.Fatalf("command %s (args %q) failed to generate coverage information", name, args)
-					return
-				}
-				coverChan <- f
+				ts.cmdExec(neg, append([]string{name}, args...))
 			}
 		}
 		return m.Run()
 	}
+	cmdName := filepath.Base(os.Args[0])
+	if runtime.GOOS == "windows" {
+		cmdName = strings.TrimSuffix(cmdName, ".exe")
+	}
 	mainf := commands[cmdName]
 	if mainf == nil {
 		log.Printf("unknown command name %q", cmdName)
 		return 2
 	}
 	// The command being registered is being invoked, so run it, then exit.
 	os.Args[0] = cmdName
-	cprof := os.Getenv("TESTSCRIPT_COVERPROFILE")
-	if cprof == "" {
+	coverdir := os.Getenv("TESTSCRIPT_COVER_DIR")
+	if coverdir == "" {
 		// No coverage, act as normal.
 		return mainf()
 	}
-	return runCoverSubcommand(cprof, mainf)
+	// For a command "foo", write ${TESTSCRIPT_COVER_DIR}/foo-${RANDOM}.
+	coverFile, err := ioutil.TempFile(coverdir, cmdName+"-*")
```

Comment:
Why do that, when we have TempFile? :) It is basically random numbers, but it also guards against (unlikely) collisions.

Reply:
Funnily enough, I had to end up doing this, because otherwise I found tests out in the wild that would error due to empty profiles to be merged. That entire problem goes away if we don't create this file ourselves, and leave it to -test.coverprofile. The new comment explains it.

Reply:
ah, cool!